src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -5594,3 +5599,47 @@ ExtractParametersFromParamList(ParamListInfo paramListInfo,
 																   parameterData->value);
 	}
 }
+
+
+/*
+ * Delegated functions which are not part of another transaction are naturally safe,
+ * but if it's part of an outer transaction (a 2PC), we should ensure that the current
+ * task will not do remote(connection) activity to ensure the transactional integrity
+ * of the current participating 2PC.
+ */
+void
+EnsureNoRemoteExecutionFromWorkers(List *taskList)
+{
+	if (taskList == NIL)
+	{
+		/* we allow executions with an empty task list (e.g. WHERE false) */
+		return;
+	}
+
+	if (!IsCitusInitiatedRemoteBackend())
+	{
+		/* we are in the client-initiated backend, remote tasks are ok */
+		return;
+	}
+
+	if (!MyBackendIsInDisributedTransaction())
+	{
+		/* if this backend is not participating in a 2PC, nothing to check */
+		return;
+	}
+
+	DistributedTransactionId *transactionId = GetCurrentDistributedTransactionId();
+	if (transactionId->initiatorNodeIdentifier == GetLocalGroupId())
+	{
+		/* Is the distributed transaction (ID) initiated on this node? */
```

Comment:
I think this condition does not make much sense to me, since separate sessions on the same node should not be allowed to perform remote operations

Reply:
We only get here when IsCitusInitiatedRemoteBackend is true (so this was a connection). The only circumstance in which we should be allowed to perform remote execution when IsCitusInitiatedRemoteBackend is true is when in a delegated procedure call. 

I guess this is very implicitly trying to check for a delegated procedure call. Can we make that check very explicit? (probably this should be in a separate PR that precedes the current one)

Reply:
Attempt at making the check very explicit: https://github.com/citusdata/citus/pull/5602

Reply:
Rebase of the above PR done.